odoo/addons/base/transpiler/js_transpiler.py

```diff
@@ -0,0 +1,467 @@
+        content += legacy_odoo_define
+    return content
+
+
+def url_to_module_path(url):
+    """
+    Odoo modules each have a name. (odoo.define("<the name>", function (require) {...});
+    It is used in to be required later. (const { something } = require("<the name>").
+    The transpiler transforms the url of the file in the project to this name.
+    It takes the module name and add a @ on the start of it, and map it to be the source of the static/src (or
+    static/tests) folder in that module.
+
+    in: web/static/src/one/two/three.js
+    out: @web/one/two/three.js
+    The module would therefore be defined and required by this path.
+
+    :param url: an url in the project
+    :return: a special path starting with @<module-name>.
+    """
+    match = re.match(r"""/?(?P<module>\w+)/([\w/]*/)?static/(?P<type>src|tests)(?P<url>/[\w/]*)""", url)
+    if match:
+        url = match["url"]
+        if url.endswith(('/index.js', '/index')):
+            url, _ = url.rsplit('/', 1)
+        if match["type"] == "src":
+            return "@%s%s" % (match['module'], url)
+        else:
+            return "@%s/../tests%s" % (match['module'], url)
+    else:
+        raise ValueError("The js file %r must be in the folder '/static/src' or '/static/test'" % url)
+
+
+def wrap_with_odoo_define(module_path, content):
+    """
+    Wraps the current content (source code) with the odoo.define call.
+    Should logically be called once all other operations have been performed.
+    """
+    return f"""odoo.define({module_path!r}, function (require) {{
+'use strict';
+let __exports = {{}};
+{content}
+return __exports;
+}});
+"""
+
+
+def convert_export_function_or_class(content, default=False):
+    """
+    Transpile functions and classes that are being exported.
+
+    .. code-block:: javascript
+
+        // before
+        export function name
+        // after
+        const name = __exports.name = function name
+
+        // before
+        export async function name
+        // after
+        const name = __exports.name = async function name
+
+        // before
+        export class name
+        // after
+        const name = __exports.name = class name
+
+    if default:
+
+        .. code-block:: javascript
+
+            // before
+            export default function name
+            // after
+            const name = __exports[Symbol.for("default")] = function name
+            // (same for the rest)
+
+    :param default: should we convert the default exports as well ?
+    """
+    pattern = r"^(?P<space>\s*)export\s+(?P<type>(async\s+)?function|class)\s+(?P<identifier>\w+)"
+    repl = r"\g<space>const \g<identifier> = __exports.\g<identifier> = \g<type> \g<identifier>"
+    if default:
+        pattern = r"^(?P<space>\s*)export\s+default\s+(?P<type>(async\s+)?function|class)\s+(?P<identifier>\w+)"
+        repl = r"""\g<space>const \g<identifier> = __exports[Symbol.for("default")] = \g<type> \g<identifier>"""
```

Comment:
This is horrible.
Why don't you make two functions (one with default, one without),
and compile those patterns outside the functions?
And BTW use the verbose syntax for regexes, it might make them readable.